test/unit/gitlab.spec.js

```diff
@@ -0,0 +1,215 @@
+      name: 'myUser',
+      access: ['$authenticated'],
+      gitlab: true
+    };
+
+    const cb: Callback = (err, data) => {
+      expect(err).toBeFalsy();
+      expect(data).toBe(true);
+      done();
+    };
+
+    verdaccioGitlab.allow_access(user, _package, cb);
+  });
+
+  test('should allow publish of package based on user group', done => {
+    const verdaccioGitlab: VerdaccioGitlab = new VerdaccioGitlab(defaultConfig, options);
+    const user: RemoteUser = {
+      real_groups: ['myGroup', 'myUser'],
+      groups: ['myGroup', 'myUser'],
+      name: 'myUser'
+    };
+    const _package: VerdaccioGitlabPackageAccess = {
+      name: '@myGroup/myPackage',
+      gitlab: true
+    };
+
+    const cb: Callback = (err, data) => {
+      expect(err).toBeFalsy();
+      expect(data).toBe(false);
+      done();
+    };
+
+    verdaccioGitlab.allow_publish(user, _package, cb);
+  });
+
+  test('should allow publish of package based on user name', done => {
+    const verdaccioGitlab: VerdaccioGitlab = new VerdaccioGitlab(defaultConfig, options);
+    const user: RemoteUser = {
+      real_groups: ['myGroup', 'myUser'],
+      groups: ['myGroup', 'myUser'],
+      name: 'myUser'
+    };
+    const _package: VerdaccioGitlabPackageAccess = {
+      name: 'myUser',
+      gitlab: true
+    };
+
+    const cb: Callback = (err, data) => {
+      expect(err).toBeFalsy();
+      expect(data).toBe(false);
+      done();
+    };
+
+    verdaccioGitlab.allow_publish(user, _package, cb);
+  });
+
+  test('should deny publish of package based on unauthenticated', done => {
+    const verdaccioGitlab: VerdaccioGitlab = new VerdaccioGitlab(defaultConfig, options);
+    const user: RemoteUser = {
+      real_groups: [],
+      groups: [],
+      name: undefined
+    };
+    const _package: VerdaccioGitlabPackageAccess = {
+      name: 'myUser',
+      gitlab: true
+    };
+
+    const cb: Callback = (err, data) => {
+      expect(err).toBeTruthy();
+      expect(data).toBeFalsy();
+      done();
+    };
+
+    verdaccioGitlab.allow_publish(user, _package, cb);
+  });
+
+  test('should deny publish of package based on group', done => {
+    const verdaccioGitlab: VerdaccioGitlab = new VerdaccioGitlab(defaultConfig, options);
+    const user: RemoteUser = {
+      real_groups: ['myGroup', 'myUser'],
+      groups: ['myGroup', 'myUser'],
+      name: 'myUser'
+    };
+    const _package: VerdaccioGitlabPackageAccess = {
+      name: '@anotherGroup/myPackage',
+      gitlab: true
+    };
+
+    const cb: Callback = (err, data) => {
+      expect(err).toBeTruthy();
+      expect(data).toBeFalsy();
+      done();
+    };
+
+    verdaccioGitlab.allow_publish(user, _package, cb);
+  });
+
+  test('should deny publish of package based on user', done => {
+    const verdaccioGitlab: VerdaccioGitlab = new VerdaccioGitlab(defaultConfig, options);
+    const user: RemoteUser = {
+      real_groups: ['myGroup', 'myUser'],
```

Comment:
again, could be nice to have a `testRemoteUser` in order to avoid duplication

Reply:
Same as above